File: web/src/layer/marker/Circle.ts
import * as L from 'leaflet';
import {Marker, MarkerJson} from './Marker';
import {MarkersLayer} from '../MarkersLayer';
import {Point} from '../../data/Point';

export class Circle extends Marker {
    constructor(layer: MarkersLayer, json: MarkerJson) {
        super(layer, json, L.circle(Point.of(json.point).toLatLng(), {
            ...json.options,
            radius: Circle.radius(json)
        }));
    }

    public override update(json: MarkerJson): void {
        super.update(json);
        (this._marker as L.Circle)
            .setLatLng(Point.of(json.point).toLatLng())
            .setRadius(Circle.radius(json));
    }

    private static radius(json: MarkerJson): number {
        return Point.pixelsToMeters((json.options as L.CircleOptions).radius ?? 10);
    }
}

File: src/layer/marker/Circle.cs
using System;
using JetBrains.Annotations;
using livemap.data;
using livemap.layer.marker.options.type;
using Newtonsoft.Json;

namespace livemap.layer.marker;

/// <summary>
/// The circle marker is used to draw circle overlays on the map
/// </summary>
/// <remarks>
/// You can make distorted circles with the <see cref="Ellipse"/>
/// </remarks>
[PublicAPI]
public class Circle : Marker {
    /// <summary>
    /// Absolute (not relative to spawn) world coordinates for the circle's center
    /// </summary>
    [JsonProperty(Order = -1)]
    public Point Point { get; set; }

    /// <inheritdoc cref="CircleOptions"/>
    [JsonProperty(Order = 10)]
    public new CircleOptions? Options { get; set; }

    /// <summary>
    /// Create a new circle at 0,0 with a random id
    /// </summary>
    public Circle() : this(Guid.NewGuid().ToString(), new Point(0, 0)) { }

    /// <summary>
    /// Create a new circle
    /// </summary>
    /// <param name="id">Unique identifying key</param>
    /// <param name="point">Absolute (not relative to spawn) world coordinates for the circle's center</param>
    /// <param name="options">Optional settings for the circle</param>
    public Circle(string id, Point point, CircleOptions? options = null) : base("circle", id) {
        Point = point;
        Options = options;
    }

    /// <inheritdoc cref="Marker.FromJson{T}"/>
    public static Circle FromJson(string json) {
        return FromJson<Circle>(json);
    }
}
